R/makeTreeSummarizedExperimentFromqiime2.R

```diff
@@ -206,55 +245,70 @@ makeTreeSummarizedExperimentFromqiime2 <- function(featureTableFile,
 #' @keywords internal
 #' @noRd
 .subset_taxa_in_feature <- function(taxa_tab, feature_tab) {
-    idx <- match(row.names(feature_tab), taxa_tab[, "Feature.ID"])
+    idx <- match(rownames(feature_tab), taxa_tab[, "Feature.ID"])
     taxa_tab <- taxa_tab[idx, , drop = FALSE]
-    row.names(taxa_tab) <- taxa_tab[, "Feature.ID"]
+    rownames(taxa_tab) <- taxa_tab[, "Feature.ID"]
 
     taxa_tab
 }
 
 #' Parse qiime2 taxa in different taxonomic levels
-#' @param taxa `tax_table` object.
+#' @param taxa_tab `data.frame` object.
 #' @param sep character string containing a regular expression, separator
-#'  between different taxonomic levels, defaults to on compatible with both
+#'  between different taxonomic levels, defaults to one compatible with both
 #'  GreenGenes and SILVA `; |;"`.
 #' @return  a `data.frame`.
 #' @keywords internal
-#' @importFrom rlang .data
-#' @importFrom tidyr separate
+#' @importFrom IRanges CharacterList IntegerList
+#' @importFrom S4Vectors DataFrame
 #' @noRd
-.parse_q2taxonomy <- function(taxa, sep = "; |;") {
-    taxa <- data.frame(taxa)
+.parse_q2taxonomy <- function(taxa_tab, sep = "; |;",
+                              removeTaxaPrefixes = FALSE, ...) {
+    if(!.is_a_bool(removeTaxaPrefixes)){
+        stop("'removeTaxaPrefixes' must be TRUE or FALSE.", call. = FALSE)
+    }
+
+    confidence <- NULL
+    featureID <- NULL
+    # make sure confidence in numeric
+    if ("Confidence" %in% colnames(taxa_tab)) {
+        confidence <- as.numeric(taxa_tab[,"Confidence"])
+    }
+    if("Feature.ID" %in% names(taxa_tab)) {
+        featureID <- taxa_tab[,"Feature.ID"]
+    }
 
     #  work with any combination of taxonomic ranks available
     all_ranks <- c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
     all_prefixes <- c("k__", "p__", "c__", "o__", "f__", "g__", "s__")
-    present_ranks <- strsplit(taxa[, "Taxon"], sep)
-    # at certain rank (e.g. genus): some taxa can not be determined which genus
-    # it assigned to
-    idx <- which.max(lengths(present_ranks))
-    present_ranks <- present_ranks[[idx]]
-    present_prefixes <- substr(present_ranks, 1, 3)
 
-    taxa <- suppressWarnings(
-        separate(
-            taxa, .data$Taxon,
-            all_ranks[match(present_prefixes, all_prefixes)],
-            sep = sep
-        )
-    )
-    taxa <- sapply(taxa, function(x) ifelse(x == "", NA_character_, x))
-    taxa <- data.frame(taxa)
-    # make sure confidence in numeric
-    if ("Confidence" %in% colnames(taxa)) {
-        taxa[["Confidence"]] <- as.numeric(taxa[["Confidence"]])
+    # split the taxa strings
+    taxa_split <- IRanges::CharacterList(strsplit(taxa_tab[,"Taxon"],sep))
+    # extract present prefixes
+    taxa_prefixes <- lapply(taxa_split, substr, 1L, 3L)
+    # match them to the order given by present_prefixes
+    taxa_prefixes_match <- lapply(taxa_prefixes, match, x = all_prefixes)
+    taxa_prefixes_match <- IRanges::IntegerList(taxa_prefixes_match)
```

Comment:
Since these two functions have been imported using `importFrom IRanges CharacterList IntegerList`, `IRanges::` is not necessary?

Reply:
no they are not